src/Core/Services/Implementations/PolicyService.cs

```diff
@@ -54,54 +55,44 @@ public class PolicyService : IPolicyService
                 var currentPolicy = await _policyRepository.GetByIdAsync(policy.Id);
                 if (!currentPolicy?.Enabled ?? true)
                 {
-                    var kickableOffenses = new List<Enums.PolicyType>() {
-                        Enums.PolicyType.TwoFactorAuthentication,
-                        Enums.PolicyType.OnlyOrg
-                    };
-
-                    if (kickableOffenses.Contains(currentPolicy.Type))
+                    Organization organization = null;
+                    var orgUsers = await _organizationUserRepository.GetManyDetailsByOrganizationAsync(
+                        policy.OrganizationId);
+                    var removableOrgUsers = orgUsers.Where(ou =>
+                        ou.Status != Enums.OrganizationUserStatusType.Invited &&
+                        ou.Type != Enums.OrganizationUserType.Owner && ou.UserId != savingUserId);
+                    switch (currentPolicy.Type)
                     {
-                        Organization organization = null;
-                        var orgUsers = await _organizationUserRepository.GetManyDetailsByOrganizationAsync(
-                            policy.OrganizationId);
-                        foreach (var orgUser in orgUsers.Where(ou =>
-                            ou.Status != Enums.OrganizationUserStatusType.Invited &&
-                            ou.Type != Enums.OrganizationUserType.Owner && ou.UserId != savingUserId))
-                        {
-                            switch (currentPolicy.Type)
+                        case Enums.PolicyType.TwoFactorAuthentication:
+                            foreach (var orgUser in removableOrgUsers)
+                            {
+                                if (!await userService.TwoFactorIsEnabledAsync(orgUser))
+                                {
+                                    organization = organization ?? await _organizationRepository.GetByIdAsync(policy.OrganizationId);
+                                    await organizationService.DeleteUserAsync(policy.OrganizationId, orgUser.Id,
+                                        savingUserId);
+                                    await _mailService.SendOrganizationUserRemovedForPolicyTwoStepEmailAsync(
+                                        organization.Name, orgUser.Email);
+                                }
+                            }
+                        break;
+                        case Enums.PolicyType.OnlyOrg:
+                            var userOrgs = await _organizationUserRepository.GetManyByManyUsersAsync(
+                                    removableOrgUsers.Select(ou => ou.UserId.Value));
+                            foreach (var orgUser in removableOrgUsers)
```

Comment:
userOrgs = all the dbo.organizationUser records for the current person the loop is iterating on
removableOrgUsers = the people we will want to remove from the organization applying the policy

One is orgUser records for an organization, the other is orgUser records for a given user inside that org

Reply:
Right, but instead of querying the database again, couldn't you just run a in-memory query on the existing `orgUsers` that you already pulled earlier?

Reply:
No, because orgUsers is filtered on an organizationId - the id for the organization applying the policy. The logic that uses userOrgs needs all organizationUser records for that user, regardless of organization. @kspearrin